### PR TITLE
Proxy locations

### PR DESCRIPTION
**Pull request by stelucz**
Actual implementation is missing possibility of defining multiple location parts.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -144,6 +144,31 @@ Simple HTTP proxy
...
+
+    nginx:
+      server:
+        site:
+          nginx_proxy_site01:
+            enabled: true
+            type: nginx_proxy_locations
+            name: site01
+            proxy:
+              locations:
```
Comment (fpytloun) [inline README.rst]: This key should be singular, `location:`

File: nginx/server/sites.sls [hunk 2]
```diff
@@ -90,6 +90,8 @@ sites-available-{{ site_name }}:
   - name: {{ server.vhost_dir }}/{{ site.type }}_{{ site.name }}.conf
   {%- if site.type == 'nginx_proxy' %}
   - source: salt://nginx/files/proxy.conf
+  {%- elif site.type == 'nginx_proxy_locations' %}
```
Comment (fpytloun) [inline nginx/server/sites.sls]: Why do you need to add new site type instead of extending `nginx_proxy`?

File: nginx/files/proxy_locations.conf [hunk 3]
```diff
@@ -0,0 +1,101 @@
+{%- set site = salt['pillar.get']('nginx:server:site:'+site_name) %}
+
+server {
+
+  {%- include "nginx/files/_name.conf" %}
+  {%- include "nginx/files/_ssl.conf" %}
+
+  {%- if site.get('ssl', {'enabled': False}).get('enabled', False) %}
+  ssl_session_cache shared:SSL:10m;
```
Comment (fpytloun) [inline nginx/files/proxy_locations.conf]: This should go to `_ssl.conf`

**Review by fpytloun — CHANGES_REQUESTED** [review]

**Comment by stelucz** [comment]
Hello @fpytloun,
thanks for comments. Intention of adding a new site type is to keep 100% backwards compatibility with existing models. If you think that it is better to extend existing `nginx_proxy` type, let me know your idea on how you would like to see it implemented. :)
Thanks!

File: nginx/files/proxy_locations.conf (hunk 3 above)
Comment (stelucz) [reply nginx/files/proxy_locations.conf]: thus `proxy.conf` should be patched too

**Comment by fpytloun** [comment]
Yes, I would prefer extending proxy type instead of defining new one. There's a lot of duplicated options.
I suggest to have something like:
```
{%- set location = {} %}
{%- if site.get('location') %}
{%- do location.update(site.location) %}
{%- else %}
{# Default / location #}
{%- do location.update({'/': {'enabled': True}})
{%- endif %}
```

in proxy.conf and iterate over it in a similar fashion as you do in proxy_location

**Comment by stelucz** [comment]
@fpytloun 
I did changes so `proxy.conf` is extended now. Locations are generated from `location:` part. 
1. If `location` part is not present in model, then `proxy` part is used.
1. If `/` location is not defined in `location` and `proxy` part is in model then `proxy` is used.
1. if `/` location and `proxy` is defined, then `/` location is used (overrides `proxy` definition)
 
I removed `ssl_session_cache`  from `_ssl_secure.conf` and `proxy.conf` as it's already defined in `_ssl.conf`

Let me know if change looks good to you or wish some other edits. Thanks!

File: README.rst [hunk 4]
```diff
@@ -144,6 +144,57 @@ Simple HTTP proxy
...
+If proxy part is defined and location is missing `/`, then proxy part is used. If `/` location is defined then it overrides proxy part. 
+
+.. code-block:: yaml
+
+    nginx:
+      server:
+        site:
+          nginx_proxy_site01:
+            enabled: true
+            type: nginx_proxy_locations
```
Comment (fpytloun) [inline README.rst]: Should be nginx_proxy

File: README.rst [hunk 5]
```diff
@@ -144,6 +144,57 @@ Simple HTTP proxy
...
+              port: 80
+
+.. code-block:: yaml
+
+    nginx:
+      server:
+        site:
+          nginx_proxy_site01:
+            enabled: true
+            type: nginx_proxy_locations
```
Comment (fpytloun) [inline README.rst]: same here

**Review by fpytloun — CHANGES_REQUESTED** [review]
Yes, it looks much better now.

File: nginx/files/proxy.conf [hunk 6]
```diff
@@ -5,28 +5,37 @@ server {
...
   underscores_in_headers on;
   {%- endif %}
 
   {%- include "nginx/files/_auth.conf" %}
 
-  location / {
-      {%- if site.proxy.upstream_proxy_pass is defined %}
-      proxy_pass {{ site.proxy.upstream_proxy_pass }};
+  {%- set location = {} %}
+  {%- if site.get('location', False) %}
```
Comment (fpytloun) [inline nginx/files/proxy.conf]: FYI: .get defaults to something evaluated as false so you don't need to explicitly use False as a default value

**Comment by stelucz** [comment]
@fpytloun Thanks, it should be fixed now.

**Comment by stelucz** [comment]
@fpytloun Hi, does it look ok?

**Comment by fpytloun** [comment]
Looks good to me now, thank you :)

**Comment by stelucz** [comment]
Thank you too! 
Have a nice week.

**Pull request merged**